controller/konnect/reconciler_controlplaneref.go

```diff
@@ -0,0 +1,239 @@
+		if e.Spec.ControlPlaneRef == nil {
+			return none
+		}
+		return mo.Some(*e.Spec.ControlPlaneRef)
+	case *configurationv1alpha1.KongUpstream:
+		if e.Spec.ControlPlaneRef == nil {
+			return none
+		}
+		return mo.Some(*e.Spec.ControlPlaneRef)
+	case *configurationv1alpha1.KongCACertificate:
+		if e.Spec.ControlPlaneRef == nil {
+			return none
+		}
+		return mo.Some(*e.Spec.ControlPlaneRef)
+	case *configurationv1alpha1.KongCertificate:
+		if e.Spec.ControlPlaneRef == nil {
+			return none
+		}
+		return mo.Some(*e.Spec.ControlPlaneRef)
+	case *configurationv1alpha1.KongVault:
+		if e.Spec.ControlPlaneRef == nil {
+			return none
+		}
+		return mo.Some(*e.Spec.ControlPlaneRef)
+	case *configurationv1alpha1.KongKey:
+		if e.Spec.ControlPlaneRef == nil {
+			return none
+		}
+		return mo.Some(*e.Spec.ControlPlaneRef)
+	case *configurationv1alpha1.KongKeySet:
+		if e.Spec.ControlPlaneRef == nil {
+			return none
+		}
+		return mo.Some(*e.Spec.ControlPlaneRef)
+	case *configurationv1alpha1.KongDataPlaneClientCertificate:
+		if e.Spec.ControlPlaneRef == nil {
+			return none
+		}
+		return mo.Some(*e.Spec.ControlPlaneRef)
+	default:
+		return none
+	}
+}
+
+// handleControlPlaneRef handles the ControlPlaneRef for the given entity.
+// It sets the owner reference to the referenced ControlPlane and updates the
+// status of the entity based on the referenced ControlPlane status.
+func handleControlPlaneRef[T constraints.SupportedKonnectEntityType, TEnt constraints.EntityType[T]](
+	ctx context.Context,
+	cl client.Client,
+	ent TEnt,
+) (ctrl.Result, error) {
+	cpRef, ok := getControlPlaneRef(ent).Get()
+	if !ok {
+		return ctrl.Result{}, nil
+	}
+
+	switch cpRef.Type {
+	case configurationv1alpha1.ControlPlaneRefKonnectNamespacedRef:
+		cp := konnectv1alpha1.KonnectGatewayControlPlane{}
+		// TODO(pmalek): handle cross namespace refs
+		nn := types.NamespacedName{
+			Name:      cpRef.KonnectNamespacedRef.Name,
+			Namespace: ent.GetNamespace(),
+		}
+		// Set namespace of control plane when it is non-empty. Only applyies for cluster scoped resources (KongVault).
+		if ent.GetNamespace() == "" && cpRef.KonnectNamespacedRef.Namespace != "" {
+			nn.Namespace = cpRef.KonnectNamespacedRef.Namespace
+		}
+		if err := cl.Get(ctx, nn, &cp); err != nil {
+			if res, errStatus := patch.StatusWithCondition(
+				ctx, cl, ent,
+				konnectv1alpha1.ControlPlaneRefValidConditionType,
+				metav1.ConditionFalse,
+				konnectv1alpha1.ControlPlaneRefReasonInvalid,
+				err.Error(),
+			); errStatus != nil || !res.IsZero() {
+				return res, errStatus
+			}
+			if k8serrors.IsNotFound(err) {
+				return ctrl.Result{}, ReferencedControlPlaneDoesNotExistError{
+					Reference: nn,
+					Err:       err,
+				}
+			}
+			return ctrl.Result{}, err
+		}
+
+		// Do not continue reconciling of the control plane has incompatible cluster type to prevent repeated failure of creation.
+		// Only CLUSTER_TYPE_CONTROL_PLANE and CLUSTER_TYPE_HYBRID are supported.
```

Comment:
We might mention here that attaching entities to CP groups is unsupported because they configuration is read only and perhaps link https://docs.konghq.com/konnect/gateway-manager/control-plane-groups/#limitations in the comment.

WDYT?

Reply:
Added comments.